luxtronik/__init__.py

```diff
@@ -103,26 +116,44 @@ def _with_lock_and_connect(self, func, *args, **kwargs):
             ret_val = func(*args, **kwargs)
             return ret_val
 
-    def read(self):
+    def read(self, data=None):
         """
-        Read data from heat pump.
-        All available data will be read from the heat pump.
+        All available data will be read from the heat pump
+        and integrated to the passed data object.
+        This data object is returned afterwards, mainly for access to a newly created.
         """
-        return self._with_lock_and_connect(self._read)
+        if data is None:
+            data = LuxtronikData()
+        return self._with_lock_and_connect(self._read, data)
 
-    def read_parameters(self):
-        """Read parameters from heat pump."""
-        return self._with_lock_and_connect(self._read_parameters)
+    def read_parameters(self, parameters=None):
+        """
+        Read parameters from heat pump and integrate them to the passed dictionary.
+        This dictionary is returned afterwards, mainly for access to a newly created.
+        """
+        if parameters is None:
+            parameters = Parameters()
+        return self._with_lock_and_connect(self._read_parameters, parameters)
 
-    def read_calculations(self):
-        """Read calculations from heat pump."""
-        return self._with_lock_and_connect(self._read_calculations)
+    def read_calculations(self, calculations=None):
+        """
+        Read calculations from heat pump and integrate them to the passed dictionary.
+        This dictionary is returned afterwards, mainly for access to a newly created.
+        """
+        if calculations is None:
+            calculations = Calculations()
+        return self._with_lock_and_connect(self._read_calculations, calculations)
 
-    def read_visibilities(self):
-        """Read visibilities from heat pump."""
-        return self._with_lock_and_connect(self._read_visibilities)
+    def read_visibilities(self, visibilities=None):
+        """
+        Read visibilities from heat pump and integrate them to the passed dictionary.
+        This dictionary is returned afterwards, mainly for access to a newly created.
+        """
+        if visibilities is None:
+            visibilities = Visibilities()
+        return self._with_lock_and_connect(self._read_visibilities, visibilities)
 
-    def write(self, parameters):
+    def write(self, parameters, data=None):
```

Comment:
I like the introduction of `_write_and_read()`. I also checked that the last release 0.3.14 did not read the parameters in `write()`. Thus, I propose that `write()` only writes the parameters and another method `write_and_read()` also reads the data after a short wait.

Reply:
Done